lang/java/reef-io/src/main/java/org/apache/reef/io/watcher/Watcher.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.io.watcher;
+
+import org.apache.reef.driver.context.ActiveContext;
+import org.apache.reef.driver.context.ClosedContext;
+import org.apache.reef.driver.context.FailedContext;
+import org.apache.reef.driver.evaluator.AllocatedEvaluator;
+import org.apache.reef.driver.evaluator.CompletedEvaluator;
+import org.apache.reef.driver.evaluator.FailedEvaluator;
+import org.apache.reef.driver.task.*;
+import org.apache.reef.io.watcher.param.EventStreams;
+import org.apache.reef.io.watcher.util.WatcherAvroUtil;
+import org.apache.reef.tang.annotations.Parameter;
+import org.apache.reef.tang.annotations.Unit;
+import org.apache.reef.wake.EventHandler;
+import org.apache.reef.wake.time.event.StartTime;
+import org.apache.reef.wake.time.event.StopTime;
+import org.apache.reef.wake.time.runtime.event.RuntimeStart;
+import org.apache.reef.wake.time.runtime.event.RuntimeStop;
+
+import javax.inject.Inject;
+import java.util.Set;
+
+/**
+ * Subscribe events and transfer them as wrapping with corresponding avro classes.
+ */
+@Unit
+public final class Watcher {
+
+  private final Set<EventStream> eventStreamSet;
+
+  @Inject
+  private Watcher(@Parameter(EventStreams.class) final Set<EventStream> eventStreamSet) {
+    this.eventStreamSet = eventStreamSet;
+  }
+
+  private void onEvent(final EventType eventType, final String jsonEncodedEvent) {
+    for (final EventStream eventStream : eventStreamSet) {
+      eventStream.onEvent(eventType, jsonEncodedEvent);
+    }
+  }
+
+  public final class DriverRuntimeStartHandler implements EventHandler<RuntimeStart> {
+
```

Comment:
Not sure if it's too verbose or not, but should we use ``private`` constructors for all these handler classes? @markusweimer what do you think?